File: attentive-android-sdk/src/test/java/com/attentive/androidsdk/AttentiveConfigTest.java
package com.attentive.androidsdk;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertNull;
import static org.junit.Assert.assertTrue;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.argThat;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.doReturn;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.mockStatic;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;
import static org.mockito.internal.verification.VerificationModeFactory.times;

import android.content.Context;
import com.attentive.androidsdk.internal.events.InfoEvent;
import com.attentive.androidsdk.internal.util.AppInfo;
import java.util.Map;
import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.mockito.MockedStatic;
import org.mockito.Mockito;

public class AttentiveConfigTest {
    private static final String DOMAIN = "DOMAINValue";
    private static final AttentiveConfig.Mode MODE = AttentiveConfig.Mode.DEBUG;
    private static final String VISITOR_ID = "visitorIdValue";
    private static final String NEW_VISITOR_ID = "newVisitorIdValue";

    private FactoryMocks factoryMocks;
    private static MockedStatic<AppInfo> mockedAppInfo;

    @Before
    public void setup() {
        factoryMocks = FactoryMocks.mockFactoryObjects();

        doReturn(VISITOR_ID).when(factoryMocks.getVisitorService()).getVisitorId();
        doReturn(NEW_VISITOR_ID).when(factoryMocks.getVisitorService()).createNewVisitorId();
        mockedAppInfo = Mockito.mockStatic(AppInfo.class);
        when(AppInfo.isDebuggable(any())).thenReturn(false);
    }

    @After
    public void cleanup() {
        factoryMocks.close();
        mockedAppInfo.close();
    }

    @Test
    public void constructor_validParams_gettersReturnConstructorParams() {
        // Arrange

        // Act
        AttentiveConfig config = new AttentiveConfig.Builder()
                .domain(DOMAIN)
                .mode(MODE)
                .context(mock(Context.class))
                .build();

        // Assert
        assertEquals(DOMAIN, config.getDomain());
        assertEquals(MODE, config.getMode());
        String visitorId = config.getUserIdentifiers().getVisitorId();
        assertTrue(visitorId != null && !visitorId.isEmpty());
        assertNotNull(config.getAttentiveApi());
        assertFalse(config.skipFatigueOnCreatives());

        verify(factoryMocks.getAttentiveApi()).sendEvent(argThat(arg -> arg instanceof InfoEvent), eq(config.getUserIdentifiers()), eq(DOMAIN));
    }

    @Test
    public void constructor_validParams_gettersReturnConstructorParams_skipFatigueAsTrue() {
        // Arrange

        // Act
        AttentiveConfig config = new AttentiveConfig.Builder()
                .domain(DOMAIN)
                .mode(MODE)
                .context(mock(Context.class))
                .skipFatigueOnCreatives(true)
                .build();

        // Assert
        assertEquals(DOMAIN, config.getDomain());
        assertEquals(MODE, config.getMode());
        String visitorId = config.getUserIdentifiers().getVisitorId();
        assertTrue(visitorId != null && !visitorId.isEmpty());
        assertNotNull(config.getAttentiveApi());
        assertTrue(config.skipFatigueOnCreatives());

        verify(factoryMocks.getAttentiveApi()).sendEvent(argThat(arg -> arg instanceof InfoEvent), eq(config.getUserIdentifiers()), eq(DOMAIN));
    }

    @Test
    public void clearUser_identifyWasPreviouslyCalledWithIdentifiers_identifiersAreCleared() {
        // Arrange
        AttentiveConfig config = new AttentiveConfig.Builder()
                .domain(DOMAIN)
                .mode(MODE)
                .context(mock(Context.class))
                .build();
        UserIdentifiers userIdentifiers = buildUserIdentifiers();
        config.identify(userIdentifiers);

        // Act
        config.clearUser();

        // Assert
        assertNull(config.getUserIdentifiers().getClientUserId());
        assertNull(config.getUserIdentifiers().getPhone());
        assertNull(config.getUserIdentifiers().getEmail());
        assertNull(config.getUserIdentifiers().getShopifyId());
        assertNull(config.getUserIdentifiers().getKlaviyoId());
        assertEquals(Map.of(), config.getUserIdentifiers().getCustomIdentifiers());
    }

    @Test
    public void clearUser_verifyNewVisitorIdCreated() {
        // Arrange
        AttentiveConfig config = new AttentiveConfig.Builder()
                .domain(DOMAIN)
                .mode(MODE)
                .context(mock(Context.class))
                .build();
        assertEquals(VISITOR_ID, config.getUserIdentifiers().getVisitorId());

        // Act
        config.clearUser();

        // Assert
        verify(factoryMocks.getVisitorService()).createNewVisitorId();
        assertEquals(NEW_VISITOR_ID, config.getUserIdentifiers().getVisitorId());
    }

    @Test
    public void identify_identifyWasPreviouslyCalledWithIdentifiers_identifiersAreUpdated() {
        // Arrange
        AttentiveConfig config = new AttentiveConfig.Builder()
                .domain(DOMAIN)
                .mode(MODE)
                .context(mock(Context.class))
                .build();
        config.identify(buildUserIdentifiers());

        // Act
        UserIdentifiers newUserIdentifiers = new UserIdentifiers.Builder()
                .withClientUserId("newClientId")
                .withPhone("+14158889999")
                .withEmail("newEmail@gmail.com")
                .withShopifyId("67890")
                .withKlaviyoId("09876")
                .withCustomIdentifiers(Map.of("key1", "newValue1", "extraKey", "extraValue"))
                .build();
        config.identify(newUserIdentifiers);

        // Assert
        assertEquals("newClientId", config.getUserIdentifiers().getClientUserId());
        assertEquals("+14158889999", config.getUserIdentifiers().getPhone());
        assertEquals("newEmail@gmail.com", config.getUserIdentifiers().getEmail());
        assertEquals("67890", config.getUserIdentifiers().getShopifyId());
        assertEquals("09876", config.getUserIdentifiers().getKlaviyoId());
        assertEquals(
                Map.of("key1", "newValue1", "key2", "value2", "extraKey", "extraValue"),
                config.getUserIdentifiers().getCustomIdentifiers());
    }

    @Test
    public void identify_withIdentifiers_sendsUserIdentifierCollectedEvent() {
        AttentiveConfig config = new AttentiveConfig.Builder()
                .domain(DOMAIN)
                .mode(MODE)
                .context(mock(Context.class))
                .build();
        UserIdentifiers userIdentifiers = buildUserIdentifiers();
        config.identify(userIdentifiers);

        verify(
                factoryMocks.getAttentiveApi(),
                times(1)
        ).sendUserIdentifiersCollectedEvent(
                eq(DOMAIN),
                eq(config.getUserIdentifiers()),
                any(AttentiveApiCallback.class));
    }

    @Test
    public void changeDomain() {
        // Arrange
        AttentiveConfig config = new AttentiveConfig.Builder()
                .domain(DOMAIN)
                .mode(MODE)
                .context(mock(Context.class))
                .build();
        UserIdentifiers userIdentifiers = buildUserIdentifiers();
        config.identify(userIdentifiers);

        // Act
        config.changeDomain("newDomain");

        // Assert
        assertEquals("newDomain", config.getDomain());
    }

    @Test
    public void changeDomain_emptyDomain() {
        // Arrange
        AttentiveConfig config = new AttentiveConfig.Builder()
                .domain(DOMAIN)
                .mode(MODE)
                .context(mock(Context.class))
                .build();
        UserIdentifiers userIdentifiers = buildUserIdentifiers();
        config.identify(userIdentifiers);

        // Act
        config.changeDomain("");

        // Assert
        assertEquals(DOMAIN, config.getDomain());

    }

    private UserIdentifiers buildUserIdentifiers() {
        return new UserIdentifiers.Builder()
                .withClientUserId("clientId")
                .withPhone("+14156667777")
                .withEmail("email@gmail.com")
                .withShopifyId("12345")
                .withKlaviyoId("54321")
                .withCustomIdentifiers(Map.of("key1", "value1", "key2", "value2"))
                .build();
    }
}

File: attentive-android-sdk/src/test/java/com/attentive/androidsdk/AttentiveConfigTest.kt
package com.attentive.androidsdk

import android.app.Application
import android.content.Context
import com.attentive.androidsdk.events.Event
import com.attentive.androidsdk.internal.events.InfoEvent
import com.attentive.androidsdk.internal.util.AppInfo
import com.attentive.androidsdk.internal.util.AppInfo.isDebuggable
import com.attentive.androidsdk.tracking.AppLaunchTracker
import org.junit.After
import org.junit.Assert
import org.junit.Before
import org.junit.Test
import org.mockito.MockedStatic
import org.mockito.Mockito
import org.mockito.internal.verification.VerificationModeFactory
import org.mockito.kotlin.any
import org.mockito.kotlin.anyOrNull
import org.mockito.kotlin.argThat
import org.mockito.kotlin.eq

class AttentiveConfigTest {
    private lateinit var factoryMocks: FactoryMocks
    private var appLaunchTracker = Mockito.mock(AppLaunchTracker::class.java)

    @Before
    fun setup() {
        factoryMocks = FactoryMocks.mockFactoryObjects()

        Mockito.doReturn(VISITOR_ID).`when`(factoryMocks.visitorService).visitorId
        Mockito.doReturn(NEW_VISITOR_ID).`when`(factoryMocks.visitorService)
            .createNewVisitorId()
        mockedAppInfo = Mockito.mockStatic(
            AppInfo::class.java
        )
        Mockito.`when`(isDebuggable(any())).thenReturn(false)
    }

    @After
    fun cleanup() {
        factoryMocks.close()
        mockedAppInfo!!.close()
    }

    @Test
    fun constructor_validParams_gettersReturnConstructorParams() {
        // Arrange

        // Act

        val config = AttentiveConfig.Builder()
            .domain(DOMAIN)
            .mode(MODE)
            .applicationContext(Mockito.mock(Application::class.java))
            .build()

        // Assert
        Assert.assertEquals(DOMAIN, config.domain)
        Assert.assertEquals(MODE, config.mode)
        val visitorId = config.userIdentifiers.visitorId
        Assert.assertTrue(!visitorId.isNullOrEmpty())
        Assert.assertNotNull(config.attentiveApi)
        Assert.assertFalse(config.skipFatigueOnCreatives())

        Mockito.verify(factoryMocks.attentiveApi).sendEvent(
            argThat{ arg: Event? -> arg is InfoEvent },
            eq(config.userIdentifiers),
            eq(
                DOMAIN
            ),
            anyOrNull()
        )
    }

    @Test
    fun constructor_validParams_gettersReturnConstructorParams_skipFatigueAsTrue() {
        // Arrange

        // Act

        val config = AttentiveConfig.Builder()
            .domain(DOMAIN)
            .mode(MODE)
            .applicationContext(Mockito.mock(Application::class.java))
            .skipFatigueOnCreatives(true)
            .build()

        // Assert
        Assert.assertEquals(DOMAIN, config.domain)
        Assert.assertEquals(MODE, config.mode)
        val visitorId = config.userIdentifiers.visitorId
        Assert.assertTrue(!visitorId.isNullOrEmpty())
        Assert.assertNotNull(config.attentiveApi)
        Assert.assertTrue(config.skipFatigueOnCreatives())

        Mockito.verify(factoryMocks.attentiveApi).sendEvent(
            argThat{ arg: Event? -> arg is InfoEvent },
            eq(config.userIdentifiers),
            eq(DOMAIN),
            anyOrNull()
        )
    }

    @Test
    fun clearUser_identifyWasPreviouslyCalledWithIdentifiers_identifiersAreCleared() {
        // Arrange
        val config = AttentiveConfig.Builder()
            .domain(DOMAIN)
            .mode(MODE)
            .applicationContext(Mockito.mock(Application::class.java))
                       .build()
        val userIdentifiers = buildUserIdentifiers()
        config.identify(userIdentifiers)

        // Act
        config.clearUser()

        // Assert
        Assert.assertNull(config.userIdentifiers.clientUserId)
        Assert.assertNull(config.userIdentifiers.phone)
        Assert.assertNull(config.userIdentifiers.email)
        Assert.assertNull(config.userIdentifiers.shopifyId)
        Assert.assertNull(config.userIdentifiers.klaviyoId)
        Assert.assertEquals(emptyMap<Any, Any>(), config.userIdentifiers.customIdentifiers)
    }

    @Test
    fun clearUser_verifyNewVisitorIdCreated() {
        // Arrange
        val config = AttentiveConfig.Builder()
            .domain(DOMAIN)
            .mode(MODE)
            .applicationContext(Mockito.mock(Application::class.java))
            .build()
        Assert.assertEquals(VISITOR_ID, config.userIdentifiers.visitorId)

        // Act
        config.clearUser()

        // Assert
        Mockito.verify(factoryMocks.visitorService).createNewVisitorId()
        Assert.assertEquals(NEW_VISITOR_ID, config.userIdentifiers.visitorId)
    }

    @Test
    fun identify_identifyWasPreviouslyCalledWithIdentifiers_identifiersAreUpdated() {
        // Arrange
        val config = AttentiveConfig.Builder()
            .domain(DOMAIN)
            .mode(MODE)
            .applicationContext(Mockito.mock(Application::class.java))
                       .build()
        config.identify(buildUserIdentifiers())

        // Act
        val newUserIdentifiers = UserIdentifiers.Builder()
            .withClientUserId("newClientId")
            .withPhone("+14158889999")
            .withEmail("newEmail@gmail.com")
            .withShopifyId("67890")
            .withKlaviyoId("09876")
            .withCustomIdentifiers(mapOf("key1" to "newValue1", "extraKey" to "extraValue"))
            .build()
        config.identify(newUserIdentifiers)

        // Assert
        Assert.assertEquals("newClientId", config.userIdentifiers.clientUserId)
        Assert.assertEquals("+14158889999", config.userIdentifiers.phone)
        Assert.assertEquals("newEmail@gmail.com", config.userIdentifiers.email)
        Assert.assertEquals("67890", config.userIdentifiers.shopifyId)
        Assert.assertEquals("09876", config.userIdentifiers.klaviyoId)
        Assert.assertEquals(
            mapOf("key1" to "newValue1", "key2" to "value2", "extraKey" to "extraValue"),
            config.userIdentifiers.customIdentifiers
        )
    }

    @Test
    fun identify_withIdentifiers_sendsUserIdentifierCollectedEvent() {
        val config = AttentiveConfig.Builder()
            .domain(DOMAIN)
            .mode(MODE)
            .applicationContext(Mockito.mock(Application::class.java))
                       .build()
        val userIdentifiers = buildUserIdentifiers()
        config.identify(userIdentifiers)

        Mockito.verify(
            factoryMocks.attentiveApi,
            VerificationModeFactory.times(1)
        ).sendUserIdentifiersCollectedEvent(
            eq(DOMAIN),
            eq(config.userIdentifiers),
            any()
        )
    }

    @Test
    fun changeDomain() {
        // Arrange
        val config = AttentiveConfig.Builder()
            .domain(DOMAIN)
            .mode(MODE)
            .applicationContext(Mockito.mock(Application::class.java))
                       .build()
        val userIdentifiers = buildUserIdentifiers()
        config.identify(userIdentifiers)

        // Act
        config.changeDomain("newDomain")

        // Assert
        Assert.assertEquals("newDomain", config.domain)
    }
    
    @Test
    fun domain_emptyString_throwsException() {
        Assert.assertThrows(IllegalArgumentException::class.java) {
            AttentiveConfig.Builder().domain("")
        }
    }

    @Test
    fun domain_containsAttnTv_throwsException() {
        Assert.assertThrows(IllegalArgumentException::class.java) {
            AttentiveConfig.Builder().domain("https://mycompany.attn.tv")
        }
    }

    @Test
    fun domain_containsColon_throwsException() {
        Assert.assertThrows(IllegalArgumentException::class.java) {
            AttentiveConfig.Builder().domain("https://mycompany")
        }
    }

    @Test
    fun domain_containsSlash_throwsException() {
        Assert.assertThrows(IllegalArgumentException::class.java) {
            AttentiveConfig.Builder().domain("mycompany/path")
        }
    }

    @Test
    fun domain_validDomain_doesNotThrow() {
        val builder = AttentiveConfig.Builder().domain("mycompany")
        // If we get here without exception, the test passes
        Assert.assertNotNull(builder)
    }

    private fun buildUserIdentifiers(): UserIdentifiers {
        return UserIdentifiers.Builder()
            .withClientUserId("clientId")
            .withPhone("+14156667777")
            .withEmail("email@gmail.com")
            .withShopifyId("12345")
            .withKlaviyoId("54321")
            .withCustomIdentifiers(mapOf("key1" to "value1", "key2" to "value2"))
            .build()
    }

    companion object {
        private const val DOMAIN = "DOMAINValue"
        private val MODE = AttentiveConfig.Mode.DEBUG
        private const val VISITOR_ID = "visitorIdValue"
        private const val NEW_VISITOR_ID = "newVisitorIdValue"

        private var mockedAppInfo: MockedStatic<AppInfo>? = null
    }
}